|script             |Description|
|-------------------|-----------|
|[`fasten_clean`](docs/bin/fasten_clean.md)     | Trims and cleans a fastq file.|
|[`fasten_convert`](docs/bin/fasten_convert.md)   | Converts between different sequence formats like fastq, sam, fasta.|
|[`fasten_straighten`](docs/bin/fasten_straighten.md)| Convert any fastq file to a standard four-line-per-entry format.|
|[`fasten_metrics`](docs/bin/fasten_metrics.md)   | Prints basic read metrics.|
|[`fasten_pe`](docs/bin/fasten_pe.md)        | Determines paired-endedness based on read IDs.|
|[`fasten_randomize`](docs/bin/fasten_randomize.md) | Randomizes reads from input |
|[`fasten_combine`](docs/bin/fasten_combine.md)   | Combines identical reads and updates quality scores.|
|[`fasten_kmer`](docs/bin/fasten_kmer.md)      | Kmer counting.|
|[`fasten_sample`](docs/bin/fasten_sample.md)    | Downsamples reads.|
|[`fasten_shuffle`](docs/bin/fasten_shuffle.md)   | Shuffles or deshuffles paired end reads.|
|[`fasten_validate`](docs/bin/fasten_validate.md)  | Validates your reads|
|[`fasten_quality_filter`](docs/bin/fasten_quality_filter.md) | Transforms nucleotides to "N" if the quality is low | |
|[`fasten_trim`](docs/bin/fasten_trim.md)      | Blunt-end trims reads | |
|[`fasten_replace`](docs/bin/fasten_replace.md)   | Find and replace using regex | |
|[`fasten_mutate`](docs/bin/fasten_mutate.md)    | introduce random mutations | |
|[`fasten_regex`](docs/bin/fasten_regex.md)     | Filter for reads using regex | |
|[`fasten_progress`](docs/bin/fasten_progress.md)  | Add progress to any place in the pipeline | |
|[`fasten_sort`](docs/bin/fasten_sort.md)  | Sort fastq entries | |
